Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp

```diff
@@ -685,6 +685,36 @@ const Object* Object::getOuterObject() const
 	return this;
 }
 
+//-------------------------------------------------------------------------------------------------
+Bool Object::isLogicallyVisible() const
+{
+	if (isLocallyViewed())
+		return true;
+
+	if (TheControlBar->isObserverControlBarOn())
```

Comment:
It would be very good to merge this with `localPlayerIsObserving`, and fix `localPlayerIsObserving` if that is not quite right yet.